test/integration-test/TestIIRDeserializer.cpp

```diff
@@ -0,0 +1,598 @@
+
+#include "dawn/Optimizer/OptimizerContext.h"
+#include "dawn/Compiler/DawnCompiler.h"
+#include "dawn/IIR/ASTStmt.h"
+#include "dawn/IIR/ASTUtil.h"
+#include "dawn/IIR/FieldAccessMetadata.h"
+#include "dawn/IIR/IIRNodeIterator.h"
+#include "dawn/IIR/InstantiationHelper.h"
+#include "dawn/IIR/StencilInstantiation.h"
+#include "dawn/Serialization/IIRSerializer.h"
+#include "dawn/Optimizer/AccessComputation.h"
+#include "dawn/Optimizer/PassComputeStageExtents.h"
+#include "dawn/Optimizer/PassTemporaryType.h"
+#include "dawn/Optimizer/PassSetStageName.h"
+#include "dawn/Optimizer/StatementMapper.h"
+#include "dawn/SIR/ASTFwd.h"
+#include "dawn/SIR/SIR.h"
+#include "dawn/Support/Logging.h"
+#include "dawn/Support/STLExtras.h"
+#include "dawn/IIR/IIR.h"
+#include "dawn/IIR/StencilMetaInformation.h"
+
+#include <stack>
+#include <memory>
+#include <string>
+#include <unistd.h>
+#include <gtest/gtest.h>
+
+using namespace dawn;
+
+namespace {
+
+void compareIIRstructures(iir::IIR* lhs, iir::IIR* rhs) {
+  ASSERT_TRUE(lhs->checkTreeConsistency());
+  ASSERT_TRUE(rhs->checkTreeConsistency());
+  // checking the stencils
+  ASSERT_TRUE(lhs->getChildren().size() == rhs->getChildren().size());
+  for(int stencils = 0, size = lhs->getChildren().size(); stencils < size; ++stencils) {
+    const auto& lhsStencil = lhs->getChild(stencils);
+    const auto& rhsStencil = rhs->getChild(stencils);
+    ASSERT_TRUE(lhsStencil->getStencilAttributes() == rhsStencil->getStencilAttributes());    
+    ASSERT_TRUE(lhsStencil->getStencilID() == rhsStencil->getStencilID());
+
+    // checking each of the multistages
+    ASSERT_TRUE(lhsStencil->getChildren().size() == rhsStencil->getChildren().size());
+    for(int mssidx = 0, mssSize = lhsStencil->getChildren().size(); mssidx < mssSize; ++mssidx) {
+      const auto& lhsMSS = lhsStencil->getChild(mssidx);
+      const auto& rhsMSS = rhsStencil->getChild(mssidx);
+      ASSERT_TRUE(lhsMSS->getLoopOrder() == rhsMSS->getLoopOrder());     
+      ASSERT_TRUE(lhsMSS->getID() == rhsMSS->getID());
+
+      // checking each of the stages
+      ASSERT_TRUE(lhsMSS->getChildren().size() == rhsMSS->getChildren().size());
+      for(int stageidx = 0, stageSize = lhsMSS->getChildren().size(); stageidx < stageSize;
+          ++stageidx) {
+        const auto& lhsStage = lhsMSS->getChild(stageidx);
+        const auto& rhsStage = rhsMSS->getChild(stageidx);
+        ASSERT_TRUE(lhsStage->getStageID() == rhsStage->getStageID());
+
+        // checking each of the doMethods
+        ASSERT_TRUE(lhsStage->getChildren().size() == rhsStage->getChildren().size());
+        for(int doMethodidx = 0, doMethodSize = lhsStage->getChildren().size();
+            doMethodidx < doMethodSize; ++doMethodidx) {
+          const auto& lhsDoMethod = lhsStage->getChild(doMethodidx);
+          const auto& rhsDoMethod = rhsStage->getChild(doMethodidx);
+          ASSERT_TRUE(lhsDoMethod->getID() == rhsDoMethod->getID());
```

Comment:
Note that as a rule of thumb, you should use `EXPECT_*` if the test can continue, and `ASSERT_*` if the test cannot continue. So here you should use `EXPECT_EQ`, while at a point where you compare the number of children and then loop over the number of children, you should use `ASSERT_EQ`.